configcat/cpp-sdk
# Config v6

Implements config v6 support.

Features/improvements:
* [x] Fetch new config json version (`config_v6.json`) and update config model
* [x] SDK key format validation (when client is not set up to use local-only flag overrides)
* [x] AND conditions
* [x] New condition types
   * Prerequisite flag condition
   * Segment condition
* [x] Changed comparison operators
   * (NOT) CONTAINS ANY OF
* [x] New comparison operators
   * BEFORE/AFTER
   * TEXT (NOT) EQUALS (both hashed and cleartext)
   * TEXT (NOT) STARTS/ENDS WITH ANY OF (both hashed and cleartext)
   * ARRAY CONTAINS ANY OF (both hashed and cleartext)
* [x] Percentage options within targeting rules
* [x] Percentage options based on custom attributes
* [x] Custom User Object attribute handling (don't force string values + autoconversions under the hood)
* [x] User Object attribute/comparison value trimming
* [x] Comparison value hashing changes (SHA256, salting)
* [x] Flag overrides vs. config json salt and segments
* [x] `EvaluationDetails.matchedTargetingRule`/`matchedPercentageOption` properties (rename + set combinations correctly)
* [x] Update `ConfigCatClient.getKeyAndValue`

Tests: 
* [x] Update cache key/payload tests
* [x] Evaluation matrix tests
  * Config v1 features (both V1 & V2)
  * Config v2 features
* [x] Evaluation log tests
* [x] SDK key format validation
* [x] Custom User Object attribute handling (don't force string values + autoconversions under the hood)
* [x] Flag overrides vs. prerequisite flags (circular dependency, comparison value type mismatch, dependent/prereq flag overriding)
* [x] Flag overrides vs. config json salt and segments
* [x] `EvaluationDetails.matchedTargetingRule`/`matchedPercentageOption` properties (rename + set combinations correctly)
* [x] Special character tests (https://trello.com/c/qh4jHrvm)
* [x] Comparison attribute and comparison value trimming tests (https://trello.com/c/vt8FFrSZ)
* [x] Comparison attribute conversion to canonical string representation tests (https://trello.com/c/CZgGrrcb)

Related improvements/fixes:
* [ ] ~Dynamically typed languages should log a warning when default value doesn't match the setting type (see https://trello.com/c/Le6vimGu)~
* [x] We should go to the cache in all polling modes instead of using the in memory variable (see https://trello.com/c/rreKm64A)
* [x] Improve message of log event 1103 (https://trello.com/c/Mb7rvm64)

Additional improvements:
- Improve `getValueDetails` API (see breaking changes). Also, add the `getValueDetails` counterpart of the `getValue` overload that doesn't require a default value.
- Handle unexpected errors in the `forceRefresh` and the evaluation methods (`getValue`, etc.) to stop potential unhandled exceptions thrown by the evaluation logic from bringing down the consumer's application.
- Return information about the success of the operation (`RefreshResult`) from the `forceRefresh` method.
- Add exception message to the log format of the default logger.
- When evaluation fails and there is a related exception, also expose it in `EvaluationDetails` via the new `errorException` field. Also, expose the exception to the `onError` hook callbacks.
- When `forceRefresh` fails and there is a related exception, also expose it in `RefreshResult` via the new `errorException` field. 

Additional bugfixes:
- Correct the config service logic to return the error message in case config fetching fails.

Breaking changes:

- **Change the return type of ConfigCatClient::get from C pointer to `shared_ptr`** to prevent use after free issues.
- **Change the type of  the `user` parameter of evaluation methods (`getValue` & co.) from C pointer to `shared_ptr`.** Also, change the type of the `EvaluationDetails.user` field to `shared_ptr` to prevent use after free issues.
- Change the type of the `ConfigCatUser.user` field from `std::unordered_map<std::string, std::string>` to  `std::unordered_map<std::string, ConfigCatUser::AttributeValue>` to allow other values than `std::string`.
- Make `EvaluationDetails` generic and change the return type of  `getValueDetails` overloads to return specialized `EvaluationDetails` objects that contain a value of the same type as the default value.
- Change the return type of  `getValue(const std::string&, const ConfigCatUser*)` and `getKeyAndValue` from `std::shared_ptr` to `std::optional` (for consistency and to remove unnecessary heap allocation).
- Add an extra parameter of type `std::exception_ptr` to the `onError` hook callback.
- Make `Settings` const in the parameter of the `onConfigChanged` hook callback as it is not supposed to be modified by consumers.
- Rename the `error` field to `errorMessage` in `EvaluationDetails` (and also make it optional).
- Rename the `matchedEvaluationRule` field to `matchedTargetingRule` and the `matchedEvaluationPercentageRule` field to `matchedPercentageOption` in `EvaluationDetails`.
- Make the `EvaluationDetails.variationId` field optional.
- Rework config model (`Config` and related structs/enums).
- Pass the potential exception as an optional parameter to the `ILogger.log` method.
- Remove the `ConfigCatLogger`, `ConfigEntry` and `SettingResult` types from the public API.
- Add an abstract `getBehavior` method to the `OverrideDataSource` class.
- Use `int32_t` instead of `int` for integer setting values for better portability (on most platforms `int32_t` is just an alias of `int`).

### Requirement checklist (only if applicable)
